### PR TITLE
Rsa key encryption

### PR DESCRIPTION
**Pull request by nickrmc83**
These changes add JWE with RSA encryption of content encryption keys. Additionally this change includes an outline of an HSM-backed version of the aforementioned capability.

- There's a fix to direct encryption mode. Previously the way the protected JWE header was processed was wrong. Whilst this did not include any obvious security floor I've added fixes that means Known Answer tests should now be able to be written with confidence. This change is not backwards compatible. Any previously encrypted blobs would fail decryption after this change. It would be wise for anyone to re-encrypt such blobs.

### CODE REVIEW

File: hsm/asymmetric_decryption_key.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+func (a *AsymmetricDecryptionKey) Kid() string {
+	return a.kid
+}
+
+// Certificates associated x509 certificates.
+func (a *AsymmetricDecryptionKey) Certificates() []*x509.Certificate {
+	// TODO: lookup certificates
+	cert, err := a.ctx.FindCertificate([]byte(a.kid), nil, nil)
+	if err != nil {
+		// :thinking:
```
Comment (nickrmc83) [inline hsm/asymmetric_decryption_key.go]: We probably need to break the `CertifiableKey` interface to propagate the error.

Comment (solcates) [reply hsm/asymmetric_decryption_key.go]: The introduction of a southbound API returning an error to .Certificates() can indeed break it.  As this is crypto11/HSM specific piece of code, could we have it be a panic(err), and mark this for a breaking change in the next major version?

Comment (nickrmc83) [reply hsm/asymmetric_decryption_key.go]: Done in f2d9de4334f85808a03d3e552fe5e801d9cc323d

**Comment by nickrmc83** [comment]
@solcates it's probably worth doing some tidy up over the next few weeks to interfaces then tagging this repo with `v1.0.0` for release.

**Pull request merged**